extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,276 @@
+      const parsedFilter = this.parseFilter_(filters[i].trim());
+      if (parsedFilter) {
+        result = result.then(value => {
+          parsedFilter.args.unshift(value);
+          return parsedFilter.fn.apply(null, parsedFilter.args);
+        });
+      }
+    }
+    return result;
+  }
+
+  /**
+   * @param {!string} template The template to expand.
+   * @param {!JSONType} trigger The object to use for variable value lookups.
+   * @param {!Object=} opt_event Object with details about the event.
+   * @param {number=} opt_iterations Number of recursive expansions to perform.
+   *    Defaults to 2 substitutions.
+   * @param {boolean=} opt_noEncode Used to determine if the vars should be
+   *    encoded or not. Defaults to false.
+   * @return {!Promise<!string>} The expanded string.
+   */
+  expandTemplate(template, trigger, config, opt_event, opt_iterations,
+      opt_noEncode) {
+    opt_iterations = opt_iterations === undefined ? 2 : opt_iterations;
+    if (opt_iterations < 0) {
+      user().error(TAG, 'Maximum depth reached while expanding variables. ' +
+          'Please ensure that the variables are not recursive.');
+      return Promise.resolve(template);
+    }
+
+    let replacementPromise;
+    let replacement = template.replace(/\${([^}]*)}/g, (match, key) => {
+      // TODO (avimehta): The parsing for variables breaks when `|` is used as
+      // something other than the filter separator. A full-fledged parser would
+      // be needed to fix this.
+
+      const tokens = key.split('|');
+      if (!tokens[0]) {
+        return Promise.resolve('');
+      }
+
+      const {name, argList} = this.getNameArgs_(tokens[0].trim());
+      // Precedence is opt_event.vars > trigger.vars > config.vars.
+      const raw = (opt_event && opt_event['vars'] && opt_event['vars'][name]) ||
+          (trigger['vars'] && trigger['vars'][name]) ||
+          (config['vars'] && config['vars'][name]) ||
+          '';
+
+      let p;
+      if (typeof raw == 'string') {
+        p = this.expandTemplate(raw, trigger, config,
+            opt_event, opt_iterations - 1, opt_noEncode);
+      } else {
+        // Values can also be arrays and objects. Don't expand them.
+        p = Promise.resolve(raw);
+      }
+
+      p = p.then(expandedValue =>
+          // First apply filters
+          this.applyFilters_(expandedValue, tokens.slice(1)))
+        .then(finalRawValue => {
+          // Then encode the value
+          const val = opt_noEncode
+              ? finalRawValue
+              : this.encodeVars(finalRawValue, name);
+          return val ? val + argList : val;
+        })
+        .then(encodedValue => {
+          // Replace it in the string
+          replacement = replacement.replace(match, encodedValue);
+        });
+
+      // Queue current replacement promise after the last replacement.
+      if (replacementPromise) {
+        replacementPromise = replacementPromise.then(() => p);
```

Comment:
We could avoid a lot of promise allocations if we use an array and `Promise.all` instead.